File: Core/Queries/DTOs/ConfigurationDto.cs
using FinanceTracker.Core.Entities;

namespace FinanceTracker.Core.Queries.DTOs;

public record ConfigurationDto(
    IReadOnlyCollection<GroupTypeConfigDto> GroupTypes,
    IReadOnlyCollection<OrderableEntityDto> PhysicalAllocations
);

public record GroupTypeConfigDto(
    Guid Key,
    string Name,
    int DisplaySequence,
    string Icon,
    bool ShowScore,
    IReadOnlyCollection<GroupConfigDto> Groups
) : OrderableEntityDto(
    Key: Key,
    Name: Name, 
    DisplaySequence: DisplaySequence
);

public record GroupConfigDto(
    Guid Key,
    string Name,
    int DisplaySequence,
    bool ShowTargets,
    Guid GroupTypeId,
    IReadOnlyCollection<ComponentConfigDto> Components
) : OrderableEntityDto(
    Key: Key,
    Name: Name, 
    DisplaySequence: DisplaySequence
);

public record ComponentConfigDto(
    Guid Key,
    string Name,
    int DisplaySequence,
    Guid GroupId,
    Guid? DefaultPhysicalAllocationId = null
) : OrderableEntityDto(
    Key: Key,
    Name: Name,
    DisplaySequence: DisplaySequence
)
{
    public static ComponentConfigDto FromComponent(Component component) =>
        new ComponentConfigDto(
            Key: component.Id,
            Name: component.Name,
            DisplaySequence: component.DisplaySequence,
            GroupId: component.GroupId,
            DefaultPhysicalAllocationId: component.DefaultPhysicalAllocationId
        );
    
    public Component ToComponent() =>
        new()
        {
            Id = Key,
            Name = Name,
            DisplaySequence = DisplaySequence,
            GroupId = GroupId,
            DefaultPhysicalAllocationId = DefaultPhysicalAllocationId
        };  
}

File: Web/ClientApp/src/api/configuration/DTOs/ConfigurationDto.ts
export type ConfigurationDto = {
    groupTypes: GroupTypeConfigDto[],
    physicalAllocations: OrderableEntityDto[]
}

export type GroupTypeConfigDto = OrderableEntityDto & {
    icon: string,
    groups: GroupConfigDto[]
}

export type GroupConfigDto = OrderableEntityDto & {
    showTargets: boolean,
    groupTypeId: string,
    components: ComponentConfigDto[]
}

export type ComponentConfigDto = OrderableEntityDto & {
    groupId: string,
    defaultPhysicalAllocationId: string | undefined,
}

export type OrderableEntityDto = {
    key: string,
    name: string,
    displaySequence: number
}

// ---

export type WalletDataDto = OrderableEntityDto & {
    components: WalletComponentDataDto[]
}

export type WalletComponentDataDto = OrderableEntityDto & {
    defaultPhysicalAllocationId: string | undefined
};